src/defineEnvironment.js

```diff
@@ -0,0 +1,117 @@
+// @flow
+import {groupTypes as htmlGroupTypes} from "./buildHTML";
+import {groupTypes as mathmlGroupTypes} from "./buildMathML";
+
+import Options from "./Options";
+import ParseNode from "./ParseNode";
+
+import type Parser from "./Parser";
+import type {ArgType, Mode} from "./types";
+
+/**
+ * The context contains the following properties:
+ *  - mode: current parsing mode.
+ *  - envName: the name of the environment, one of the listed names.
+ *  - parser: the parser object.
+ */
+type EnvContext = {|
+    mode: Mode,
+    envName: string,
+    parser: Parser,
+|};
+
+/**
+ * The handler function receives two arguments
+ *  - context: information and references provided by the parser
+ *  - args: an array of arguments passed to \begin{name}
+ */
+type EnvHandler = (context: EnvContext, args: ParseNode[]) => ParseNode;
+
+/**
+ *  - numArgs: (default 0) The number of arguments after the \begin{name} function.
+ *  - argTypes: (optional) Just like for a function
+ *  - allowedInText: (default false) Whether or not the environment is allowed
+ *                   inside text mode (not enforced yet).
+ *  - numOptionalArgs: (default 0) Just like for a function
+ */
+type EnvProps = {
+    numArgs?: number,
+    argTypes?: ArgType[],
+    allowedInText?: boolean,
+    numOptionalArgs?: number,
+};
+
+/**
+ * Final enviornment spec for use at parse time.
+ * This is almost identical to `EnvDefSpec`, except it
+ * 1. includes the function handler
+ * 2. requires all arguments except argType
+ * It is generated by `defineEnvironment()` below.
+ */
+type EnvSpec = {|
```

Comment:
I tightened up some of these types by updating them to use exact object types.

Reply:
Meta-comment on GitHub reviews vs. Phabricator: I'm finding it frustrating that comments need to pin to a single line.